Use address config domain to load location metadata

jecihjoy opened the pull request.

ojwanganto commented:
Unfortunately, I don't have a good experience with the initializer module. I think @makombe @donaldkibet are better placed to review

donaldkibet commented:
@ojwanganto this seems okay. We just need to test it out

ibacher commented:
This is where Iniz's Validator can be useful for verifying that changes to the metadata still successfully load. In the RefApp, we use [this GitHub Action](https://github.com/openmrs/openmrs-distro-referenceapplication/blob/main/.github/workflows/build-test.yml) on PR to validate metadata. It uses [this Maven profile](https://github.com/openmrs/openmrs-distro-referenceapplication/blob/ee9e97e84258ab1545905e9d3db22c69edefe240/distro/pom.xml#L294) to trigger the Iniz validator, which spins up an empty database and loads the configuration into it.

Off the top of my head, this PR looks fine, though I'm a little concerned about how `<value>` will handle the XML, since it's mostly used to load much simpler strings.

rbuisson commented:
Probably won't work indeed. Take a look at this:
https://github.com/openmrs/ozone-distro-cambodia/blob/main/configs/openmrs_config/globalproperties/address-template.xml
The value inside `<value>` must be escaped.

+1 to use the `validate` profile of Iniz.

jecihjoy commented:
Thank you for your reviews above, from your feedbacks
1. I have updated loading of location metadata from GP domain to addressHierachy initializer domain
2. Integrated the initializer validator which runs when a PR is raised against `main` branch

While integrating the validator, I changed the initializer folder name from `configs` to `configuration` because of this error

```
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  14.606 s
[INFO] Finished at: 2024-01-05T19:15:32Z
[INFO] ------------------------------------------------------------------------
Error:  Failed to execute goal org.openmrs.maven.plugins:openmrs-packager-maven-plugin:1.7.0:validate-configurations (validate-configurations) on project ozone-kenya: /home/runner/work/kenyahmis/kenyahmis/configuration does not point to a valid directory. -> [Help 1]
Error:  
Error:  To see the full stack trace of the errors, re-run Maven with the -e switch.
Error:  Re-run Maven using the -X switch to enable full debug logging.
Error:  
Error:  For more information about the errors and possible solutions, please read the following articles:
Error:  [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoExecutionException
Error: Process completed with exit code
```

Also when GA runs, the build success with the error below, am not sure whether that means the metadata test was a success or a failuire

```
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  07:34 min
[INFO] Finished at: 2024-01-05T20:10:35Z
[INFO] ------------------------------------------------------------------------
[INFO] Closing org.springframework.context.support.GenericApplicationContext@6d9327d2: startup date [Fri Jan 05 20:08:04 UTC 2024]; root of context hierarchy
Error: Exception in thread "Thread-3" java.lang.NoClassDefFoundError: org/testcontainers/utility/PathUtils
	at org.testcontainers.utility.MountableFile.lambda$deleteOnExit$0(MountableFile.java:292)
	at java.lang.Thread.run(Thread.java:750)
Caused by: java.lang.ClassNotFoundException: org.testcontainers.utility.PathUtils
	at org.codehaus.plexus.classworlds.strategy.SelfFirstStrategy.loadClass(SelfFirstStrategy.java:50)
	at org.codehaus.plexus.classworlds.realm.ClassRealm.unsynchronizedLoadClass(ClassRealm.java:271)
	at org.codehaus.plexus.classworlds.realm.ClassRealm.loadClass(ClassRealm.java:247)
	at org.codehaus.plexus.classworlds.realm.ClassRealm.loadClass(ClassRealm.java:239)
	... 2 more
```

Apart from integrating with with iniz validator, I would wish to test loading of metadata locally since I was able to run the pipeline. Docker creates two db containers, mariadb and postgre, which one has openmrs and what are the mysql user credentials

ibacher commented:
> While integrating the validator, I changed the initializer folder name from configs to configuration because of this error

This isn't quite the right fix. That directory is not a valid Initializer configuration, so renaming it is just broken in a different way. We need to point the Validator to the `configs/initializer_config` directory. This can be done by passing that as the `sourceDir` argument:

```xml
<profile>
      <id>validator</id>
      <build>
        <plugins>
          <plugin>
            <groupId>org.openmrs.maven.plugins</groupId>
            <artifactId>openmrs-packager-maven-plugin</artifactId>
            <executions>
              <execution>
                <id>validate-configurations</id>
                <phase>validate</phase>
                <goals>
                  <goal>validate-configurations</goal>
                </goals>
                <configuration>
                   <sourceDir>${project.basedir}/configs/initializer_config</sourceDir>
                </configuration>
              </execution>
            </executions>
          </plugin>
        </plugins>
      </build>
    </profile>
```

> when GA runs, the build success with the error below, am not sure whether that means the metadata test was a success or a failuire

The build is a success, (although no metadata is loaded). That exception occurs in part of the TestContainer clean-up; it shouldn't be happening, but I'd probably have to look closer at the logs to have an idea why.

>  I would wish to test loading of metadata locally since I was able to run the pipeline. Docker creates two db containers, mariadb and postgre, which one has openmrs and what are the mysql user credentials

MariaDB has the OpenMRS database. The credentials for all the services are the defaults, laid out in [this file](https://github.com/ozone-his/ozone-docker/blob/main/.env). Of course, eventually we will swap those out with something real.

ibacher commented on configs/openmrs/initializer_config/addresshierarchy/addressConfiguration.xml:
```diff
@@ -0,0 +1,32 @@
+		<entryDelimiter>,</entryDelimiter>
+		<identifierDelimiter>^</identifierDelimiter>
+	</addressHierarchyFile>
+</addressConfiguration>
```
```suggestion
</addressConfiguration>

```
